cloudinit/config/cc_write_files.py

```diff
@@ -118,6 +135,40 @@ def decode_perms(perm, default):
         return default
 
 
+def read_url_or_decode(source, ssl_details, content, encoding):
+    contents = None
+    # Fetch file content from source URL, if provided
+    url = None if source == None else source.get("uri", None)
+    use_url = bool(url)
+    if use_url:
+        try:
+            # NOTE: These retry parameters are arbitrarily chosen defaults.
+            # They have no significance, and may be changed if appropriate
+            contents = url_helper.read_file_or_url(
```

Comment:
Nit: `content` and `contents` being used in various places in the same function can be a little messy. Maybe make this `uri_contents` or something similar?